controllers/virtualmachine/volume/volume_controller.go

```diff
@@ -785,6 +793,110 @@ func (r *Reconciler) createCNSAttachment(
+	volume vmopv1.VirtualMachineVolume,
+) error {
+
+	if volume.PersistentVolumeClaim == nil || volume.PersistentVolumeClaim.InstanceVolumeClaim != nil {
+		return nil
+	}
+
+	pvc := corev1.PersistentVolumeClaim{}
+	pvcKey := client.ObjectKey{
+		Namespace: ctx.VM.Namespace,
+		Name:      volume.PersistentVolumeClaim.ClaimName,
+	}
+
+	if err := r.Get(ctx, pvcKey, &pvc); err != nil {
+		return fmt.Errorf("cannot get PVC: %w", err)
+	}
+
+	if pvc.Status.Phase == corev1.ClaimBound {
+		// Regardless of the StorageClass binding mode there is nothing to done if already bound.
+		return nil
+	}
+
+	if pvc.Annotations[constants.KubernetesSelectedNodeAnnotationKey] != "" {
+		// Once set, this annotation cannot really be changed so just keep going.
+		// TBD: Make check if the Node still exists, and is in the VM's Zone?
+		return nil
+	}
+
+	scName := pvc.Spec.StorageClassName
+	if scName == nil || *scName == "" {
+		// While we have the DefaultStorageClass feature gate enabled, we don't mark an
+		// StorageClass with storageclass.kubernetes.io/is-default-class, so don't bother
+		// trying to resolve that here now and just continue the prior flow.
+		return nil
+	}
+
+	sc := &storagev1.StorageClass{}
+	if err := r.Get(ctx, client.ObjectKey{Name: *scName}, sc); err != nil {
+		return fmt.Errorf("cannot get StorageClass for PVC %s: %w", pvc.Name, err)
+	}
+
+	if mode := sc.VolumeBindingMode; mode == nil || *mode != storagev1.VolumeBindingWaitForFirstConsumer {
+		return nil
+	}
+
+	// We just need the name of any Node that is in this VM's Zone.
+	nodeName, err := getANodeNameInVMZone(ctx, r.Client)
+	if err != nil {
+		return fmt.Errorf("cannot find node for VM: %w", err)
+	}
+
+	if pvc.Annotations == nil {
+		pvc.Annotations = map[string]string{}
+	}
+	pvc.Annotations[constants.KubernetesSelectedNodeAnnotationKey] = nodeName
+
+	if err := r.Client.Update(ctx, &pvc); err != nil {
+		return fmt.Errorf("cannot update PVC to add selected-node annotation: %w", err)
+	}
+
+	return nil
+}
+
+// getANodeNameInVMZone returns one of the Nodes' name that is in the same
+// Zone as the VM.
+func getANodeNameInVMZone(
+	ctx *pkgctx.VolumeContext,
+	k8sClient client.Client) (string, error) {
+
+	zoneName := ctx.VM.Status.Zone
+	if zoneName == "" {
+		// Fallback to the label value if Status hasn't been updated yet.
+		zoneName = ctx.VM.Labels[topology.KubernetesTopologyZoneLabelKey]
+		if zoneName == "" {
+			return "", fmt.Errorf("VM does not have Zone set")
+		}
+	}
+
+	nodes := &corev1.NodeList{}
+	if err := k8sClient.List(
+		ctx,
+		nodes,
+		client.MatchingLabels{topology.KubernetesTopologyZoneLabelKey: zoneName}); err != nil {
+		return "", err
+	}
+
+	if len(nodes.Items) == 0 {
+		return "", fmt.Errorf("no Nodes for zone %s", zoneName)
+	}
+
+	for i := range nodes.Items {
+		if nodes.Items[i].DeletionTimestamp.IsZero() {
+			return nodes.Items[i].Name, nil
+		}
+	}
+
+	// Otherwise just return the first one.
+	return nodes.Items[0].Name, nil
```

Comment:
Is there a point at all of returning a node that is being deleted?  Can a PVC be even bound to an object on such a node?

Reply:
Nothing is being done to the specific node here - we just need a node that's in the zone. Might as well try avoid nodes that might go away before CSI processes the PVC but this specifically is an not expected edge case since the VM exists.